javalib/src/main/scala/java/net/URI.scala

```diff
@@ -601,39 +614,91 @@ object URI {
 
   // Quote helpers
 
-  private val quoteChar: js.Function1[String, String] = { (str: String) =>
-    require(str.length == 1)
+  private def decodeComponent(str: String): String = {
```

Comment:
Maybe we should fast-path the case where `str` doesn't contain any `%`. Because the charset machinery is quite heavy and rarely needed. Something like:
```scala
if (str.forall(_ != '%')) str
else {
  ...
}
```